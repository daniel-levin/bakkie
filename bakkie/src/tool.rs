use bakkie_schema::{CallToolRequestParams, CallToolResult};
use schemars::Schema;
use serde::Serialize;
use serde_json::{Map, Value};
use std::{collections::HashMap, pin::Pin};
use thiserror::Error;

#[derive(Debug)]
pub struct ToolOutput {}

#[derive(Debug, Error)]
pub enum ToolError {}

#[derive(Debug, Serialize)]
pub struct Tool {
    pub name: String,
    pub title: String,
    pub description: String,

    #[serde(rename = "inputSchema")]
    pub input_schema: crate::schemars::Schema,

    #[serde(skip)]
    pub construct_fn: ConstructTool,
}

pub type ConstructTool = fn(Map<String, Value>) -> ToolFut;

pub type ToolFut = Pin<Box<dyn Future<Output = Result<ToolOutput, ToolError>> + Send>>;

#[derive(Debug, Default)]
pub(crate) struct Tools {
    pub registry: HashMap<String, Tool>,
}

impl Tools {
    /// Inject the request body into a future that may be awaited, if such a tool exists.
    pub fn prepare_future(&self, request: CallToolRequestParams) -> Option<ToolFut> {
        if let Some(tool) = self.registry.get(&request.name) {
            let fut = (tool.construct_fn)(request.arguments);
            Some(fut)
        } else {
            None
        }
    }
}
