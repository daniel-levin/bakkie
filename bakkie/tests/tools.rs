use bakkie::tool::Tool;
use schemars::{JsonSchema, schema_for};
use serde::Serialize;

#[test]
fn serialize() {
    #[derive(Serialize, JsonSchema)]
    struct City {
        location: String,
    }

    let s = schema_for!(City);

    let t = Tool {
        name: "get_weather".into(),
        title: "Weather Information Provider".into(),
        description: "Get current weather information".into(),
        input_schema: s,
        construct_fn: |x| Box::pin(async { unimplemented!() }),
    };

    insta::assert_json_snapshot!(t);
}
