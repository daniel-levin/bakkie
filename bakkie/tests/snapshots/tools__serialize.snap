---
source: bakkie/tests/tools.rs
expression: t
---
{
  "name": "get_weather",
  "title": "Weather Information Provider",
  "description": "Get current weather information",
  "inputSchema": {
    "$schema": "https://json-schema.org/draft/2020-12/schema",
    "title": "City",
    "type": "object",
    "properties": {
      "location": {
        "type": "string"
      }
    },
    "required": [
      "location"
    ]
  }
}
